### PR TITLE
REPO-3616: Enable collectors to schedule non locking jobs

### PR DESCRIPTION
**Pull request by eknizat**
* Extend design to allow for addition of custom jobs via injecting into collector beans
* Add new job which allows collectors to collect data across all servers in a cluster without locking

### CODE REVIEW

**Comment by ancutam** [comment]
Looks good! 
There is some duplication between `HeartBeatJob` and `NonLockingHeartbeatJob`, it would be good to extract the common parts to an abstract class, easier to see what is the difference between the two types of jobs.

**Comment by derekhulley** [comment]
Looks good.
Some style comments.
Please remove redundant empty lines and @param declarations that do not add anything javadoc would not do anyway.

File: src/main/java/org/alfresco/heartbeat/HBDataCollectorServiceImpl.java [hunk 1]
```diff
@@ -282,16 +185,16 @@ public synchronized void onLicenseFail()
 
     private void restartAllCollectorSchedules()
     {
-        for(HBBaseDataCollector collector : collectors)
+        for( Map.Entry<String, HBBaseDataCollector>  entry : collectors.entrySet() )
```
Comment (ancutam) [inline src/main/java/org/alfresco/heartbeat/HBDataCollectorServiceImpl.java]: You can iterate here through the `collectors.values()` collection.

File: src/test/java/org/alfresco/heartbeat/HBDataCollectorServiceImplTest.java [hunk 2]
```diff
@@ -58,17 +52,14 @@
     private final String INVALID_CRON = "0 no no * * ?";
```
Comment (ancutam) [inline src/test/java/org/alfresco/heartbeat/HBDataCollectorServiceImplTest.java]: not used anymore

File: src/test/java/org/alfresco/heartbeat/jobs/QuartzJobSchedulerTest.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    {
+        final String cron1 = "0 0/1 * * * ?";
+        final String cron2 = "0 0/2 * * * ?";
+        final String cron3 = "0 0/3 * * * ?";
+
+        final HBBaseDataCollector c1 = new SimpleHBDataCollector("c1", cron1);
+        final HBBaseDataCollector c2 = new SimpleHBDataCollector("c2", cron2);
+        final HBBaseDataCollector c3 = new SimpleHBDataCollector("c3", cron3);
+
+        final String triggerName3 = "heartbeat-" + c3.getCollectorId() + "-Trigger";
```
Comment (ancutam) [inline src/test/java/org/alfresco/heartbeat/jobs/QuartzJobSchedulerTest.java]: Not used, please remove.

**Comment by ancutam** [comment]
Looks very good, just few minor changes.

File: src/main/java/org/alfresco/heartbeat/HBDataCollectorServiceImpl.java (hunk 1 above)
Comment (eknizat) [reply src/main/java/org/alfresco/heartbeat/HBDataCollectorServiceImpl.java]: Good idea, I'v made the change

File: src/main/java/org/alfresco/heartbeat/HBBaseDataCollector.java [hunk 4]
```diff
@@ -100,6 +109,25 @@ public void setHbDataCollectorService(HBDataCollectorService hbDataCollectorServ
...
+        this.hbJobScheduler = hbJobScheduler;
+    }
+
+    /**
+     *
+     * @return JobScheduler used to schedule jobs for this collector.
+     */
+    public HeartBeatJobScheduler getHbJobScheduler()
+    {
+        PropertyCheck.mandatory(this, "hbJobScheduler", hbJobScheduler);
```
[the remark this answers is not in the document]
Comment (eknizat) [reply src/main/java/org/alfresco/heartbeat/HBBaseDataCollector.java]: Thanks for the review Derek. I'v moved this into the constructor  and the property check is done there as well.

**Pull request merged**